Framework/PythonInterface/plugins/algorithms/AlignComponents.py

```diff
@@ -623,30 +627,37 @@ def _minimisation_func(self, x_0, wks_name, component, firstIndex, lastIndex):
         # calculate the fractional peak center deviations, then sum their absolute values
         return np.sum(np.abs((peaks_d - self.peaks_ref) / self.peaks_ref))
 
-    def _getFirstDetID(self, component):
-        """
-        recursive search to find first detID of a component
+    def _unique_name(self, component, component_info):
+        r"""Given the full name (or part of the full name) of a component, find the part
+        of the name that is unique in the whole instrument.
 
-        @param component :: reference to a detector component object
+        Example: the unique name for 'CORELLI/A row/bank1/sixteenpack' is 'bank1'. There's only
+        one 'bank1' in the whole instrument
 
-        @returns detector ID (``int``) of the first detector in the component
+        @param str component: (partial) full name of the component assembly
+        @param mantid.geometry.componentInfo component_info: object holding information for the instrument components
+        @return str: the unique name
         """
-        if component.type() == 'DetectorComponent' or component.type() == 'GridDetectorPixel':
-            return component.getID()
-        else:
-            return self._getFirstDetID(component[0])
-
-    def _getLastDetID(self, component):
+        name_parts = sorted(component.split('/'), reverse=True)
+        for part in name_parts:
+            if component_info.uniqueName(part):
+                return part
+        raise RuntimeError('Could not find a unique name for {component}')
+
+    def _firstAndLastDetID(self, component, component_info):
+        r"""first and last detector ID's in a component
+        @param str component: name of the component assembly
+        @param mantid.geometry.componentInfo component_info: object holding information for the instrument components
+        @return tuple: firt and last detector ID's
         """
-        recursive search to find last detID of a component
-
-        @param component :: reference to a detector component object
-
-        @returns detector ID (`int`) of the last detector in the component        """
-        if component.type() == 'DetectorComponent' or component.type() == 'GridDetectorPixel':
-            return component.getID()
-        else:
-            return self._getLastDetID(component[component.nelements() - 1])
+        unique_name = self._unique_name(component, component_info)
+        index = component_info.indexOfAny(unique_name)  # component-info index
```

Comment:
What happens if the name doesn't exist in the instrument?

Reply:
I see what you mean. `AlignComponents` didn't check for incorrect component names but I guess we can now `raise` some screaming to that effect here.